src/python/pants/engine/target.py

```diff
@@ -808,14 +809,81 @@ class AllTargetsRequest:
+
+    To act as a generator, a Target type should subclass this mixin and install generation `@rule`s
+    which consume a corresponding GenerateTargetsRequest subclass to produce GeneratedTargets.
+    """
+
+    # The generated Target class.
+    generated_target_cls: ClassVar[type[Target]]
+
+    # Fields which have their values copied from the generator Target to the generated Target.
+    #
+    # Must be a subset of `core_fields`.
+    copied_fields: ClassVar[Tuple[Type[Field], ...]]
+
+    # Fields which are specified to instances of the generator Target, but which are propagated
+    # to generated Targets rather than being stored on the generator Target.
+    #
+    # Must be disjoint from `core_fields`.
+    moved_fields: ClassVar[Tuple[Type[Field], ...]]
+
+
+class TargetFilesGenerator(TargetGenerator):
+    """A TargetGenerator which generates a Target per file matched by the generator.
+
+    Unlike TargetGenerator, no additional `@rules` are required to be installed, because generation
+    is implemented declaratively. But an optional `settings_request_cls` can be declared to
+    dynamically control some settings of generation.
+    """
+
+    settings_request_cls: ClassVar[type[TargetFilesGeneratorSettingsRequest] | None] = None
+
+
+@union
+class TargetFilesGeneratorSettingsRequest:
+    """An optional union to provide dynamic settings for a `TargetFilesGenerator`.
+
+    See `TargetFilesGenerator`.
+    """
+
+
+@dataclass
+class TargetFilesGeneratorSettings:
+    # Set `add_dependencies_on_all_siblings` to True so that each file-level target depends on all
+    # other generated targets from the target generator. This is useful if both are true:
+    #
+    # a) file-level targets usually need their siblings to be present to work. Most target types
+    #   (Python, Java, Shell, etc) meet this, except for `files` and `resources` which have no
+    #   concept of "imports"
+    # b) dependency inference cannot infer dependencies on sibling files.
+    #
+    # Otherwise, set `add_dependencies_on_all_siblings` to `False` so that dependencies are
+    # finer-grained.
+    add_dependencies_on_all_siblings: bool = False
+
+
+_TargetGenerator = TypeVar("_TargetGenerator", bound=TargetGenerator)
 
 
 @union
 @dataclass(frozen=True)
-class GenerateTargetsRequest(Generic[_Tgt]):
-    generate_from: ClassVar[type[_Tgt]]
-    generator: _Tgt
+class GenerateTargetsRequest(Generic[_TargetGenerator]):
+    generate_from: ClassVar[type[_TargetGenerator]]
+
+    # The TargetGenerator instance to generate targets for.
+    generator: _TargetGenerator
+    # The base Address to generate for. Note that due to parametrization, this may not
+    # always be the Address of the underlying target.
+    template_address: Address
+    # The `TargetGenerator.moved_field/copied_field` Field instances that the generator
+    # should generate targets with.
+    template: dict[str, Field] = dataclasses.field(hash=False)
```

Comment:
I'm confused what the key is?

Reply:
It's a field name. Will clarify.

I'm also confused as to why mypy hasn't complained here, because this is `dict[str, Any]` (since these are raw field values, not yet parsed).

Reply:
The key is a duplicate of `FIeld.alias`, but I imagine that's intentional for O(1) lookup rather than an O(n) scan?  (Sorry having a hard time tracking down how this is consumed)